check aggregate

danielhuppmann opened the pull request:
This PR adds a function to check whether the sub-categories of a timeseries (e.g., `foo|bar1`, `foo|bar2`) match the values of the variable of the level above (e.g., `foo`). This helps to identify reporting errors and missing entries.

gidden commented:
Some small changes requested. I think this is also a feature that would be very useful to document with examples (especially because I am not immediately sure how to use it). Perhaps it is worth starting an "Examples" page or something similar in the documentation to show how to use this function? I would imagine we will add lots more too, basically like the current plotting gallery but for other operations.

danielhuppmann commented:
@gidden, thanks for the feedback! should be good to go...

gidden commented on tests/test_core.py:
```diff
@@ -202,6 +202,30 @@ def test_validate_top_level(meta_df):
 
+def test_check_aggregate_fail(meta_df):
+    obs = meta_df.check_aggregate('Primary Energy', exclude=True)
+    print(obs)
```
rm print

gidden commented on pyam/core.py:
```diff
@@ -392,20 +393,60 @@ def validate(self, criteria={}, exclude=False):
 
             return df
 
+    def check_aggregate(self, var_total, var_sector=None, units=None,
```
a few questions on function names, feel free to disagree:

- `var_total` to `variable`
- `var_sector` to `components`

gidden commented on pyam/core.py:
```diff
@@ -392,20 +393,60 @@ def validate(self, criteria={}, exclude=False):
             return df
 
+    def check_aggregate(self, var_total, var_sector=None, units=None,
+                        exclude=False, threshold=0.1, multiplier=1):
```
Is there a use case for the multiplier?

danielhuppmann replied:
one use case in the SR1.5 context is to check that values are equal but with different signs (like CCS quantity vs. carbon removed must be equal, but one is usually reported as positive and the other as negative). Or going forward, making sure that timeseries in different units are equal...

gidden commented on pyam/core.py:
```diff
@@ -392,20 +393,60 @@ def validate(self, criteria={}, exclude=False):
+        n = len(df_total)
+
+        diff = df_total - multiplier * df_sector
+        diff = diff[abs(diff) > threshold]
```
perhaps you could use here

```
np.isclose(df_total, df_sector, **kwargs)
```

and then allow tolerances and such in the kwargs

see https://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.isclose.html

danielhuppmann commented:
thanks @gidden for the suggestion with `np.isclose()` - a lot nice!also incorporated your renaming suggestions

rebased on top of PR #55, ready for review

gidden commented:
This might be easiest to cherry pick now

```
git checkout master
git checkout -b new-branch
git cherry-pick 26d63e0..bc41dd1
```

danielhuppmann commented:
rebased on `2695bc`

danielhuppmann commented:
should be good to be merged, @gidden

gidden commented:
lgtm!
